src/workerd/io/worker.c++

```diff
@@ -2420,75 +2536,183 @@ private:
+    if (maybeLimitError == nullptr) {
+      auto limitScope = isolate.getLimitEnforcer().enterInspectorJs(*lock, maybeLimitError);
+      lock->v8Isolate->PerformMicrotaskCheckpoint();
+    } else {
+      // Oops, we already exceeded the limit, so force the microtask queue to be thrown away.
+      lock->v8Isolate->TerminateExecution();
+      lock->v8Isolate->PerformMicrotaskCheckpoint();
+    }
 
-    mutable kj::Maybe<kj::PromiseFulfillerPair<void>> paf = kj::newPromiseAndFulfiller<void>();
-    // Accessed only in notifier's owning thread.
+    KJ_IF_MAYBE(limitError, maybeLimitError) {
+      v8::HandleScope scope(lock->v8Isolate);
 
-    mutable bool inFlight = false;
-    // Is a notification already in-flight?
-  };
+      // HACK: We want to print the error, but we need a context to do that.
+      //   We don't know which contexts exist in this isolate, so I guess we have to
+      //   create one. Ugh.
+      auto dummyContext = v8::Context::New(lock->v8Isolate);
+      auto& inspector = *KJ_ASSERT_NONNULL(isolate.impl->inspector);
+      inspector.contextCreated(
+          v8_inspector::V8ContextInfo(dummyContext, 1, v8_inspector::StringView(
+              reinterpret_cast<const uint8_t*>("Worker"), 6)));
+      sendExceptionToInspector(inspector, dummyContext,
+          jsg::extractTunneledExceptionDescription(limitError->getDescription()));
+      inspector.contextDestroyed(dummyContext);
+    }
 
-  kj::Own<XThreadNotifier> outgoingQueueNotifier = kj::atomicRefcounted<XThreadNotifier>();
-  // Whenever another thread adds messages to the outgoing queue, it notifies the inspector
-  // connection thread using this.
+    if (recordedLock.checkInWithLimitEnforcer(isolate)) {
+      disconnect();
+    }
+  }
+};
 
-  kj::MutexGuarded<kj::Vector<kj::String>> outgoingQueue;
-  bool receivedClose = false;
+void Worker::Isolate::InspectorChannelImpl::pauseIncomingMessages() {
+  ioWorker.pauseIncomingMessages();
+}
 
-  volatile bool networkEnabled = false;
-  // Not under `state` lock due to lock ordering complications.
+void Worker::Isolate::InspectorChannelImpl::resumeIncomingMessages() {
+  ioWorker.resumeIncomingMessages();
+}
 
-  kj::Promise<void> sendToWebsocket(kj::ArrayPtr<kj::String> messages) {
-    if (messages.size() == 0) {
-      return kj::READY_NOW;
+bool Worker::Isolate::InspectorChannelImpl::dispatchOneMessageDuringPause() {
+  auto maybeMessage = ioWorker.waitForMessage();
+
+  // We can be paused by either hitting a debugger statement in a script or from hitting
+  // a breakpoint or someone hit break. As a result we may or may not be recursive dispatching messages.
+
+  KJ_IF_MAYBE(message, maybeMessage) {
+    auto lockedState = this->state.lockExclusive();
+    if (IoContext::hasCurrent()) {
+      // Received a message whilst script is running, probably in a breakpoint.
+      v8_inspector::V8InspectorSession& session = *lockedState->get()->session;
+      // const_cast OK because we're going to lock it
```

Comment:
I guess more accurately, const_cast OK because we already have the lock via the IoContext?

Reply:
Thanks, done.